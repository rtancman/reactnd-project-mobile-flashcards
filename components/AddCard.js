import React, { Component } from 'react'
import { View } from 'react-native'
import { connect } from 'react-redux'
import { NavigationActions } from 'react-navigation'
import { PressBtn, Grid, Title, CustomTextField, CustomSwitch } from '../utils/layout'
import { addCardToDeckFetch } from '../actions'

class AddCard extends Component {
  static navigationOptions = ({ navigation }) => {
    const { title } = navigation.state.params
    return {
      title: title
    }
  }

  state = {
    question: '',
    errorQuestion: '',
    answer: '',
    errorAnswer: '',
    correct: false,
  }

  onSubmit(deckId) {
    const { question, answer, correct } = this.state

    if( question === '' && answer === '' ) {
      this.setState({
        errorQuestion: 'this field is required',
        errorAnswer: 'this field is required',
      })
    } else if( question === '' ) {
      this.setState({ errorQuestion: 'this field is required'})
    } else if( answer === '' ) {
      this.setState({ errorAnswer: 'this field is required'})
    } else {
      this.props.dispatch(addCardToDeckFetch(deckId, {question, answer, correct}))
        .then((card) => {
          this.setState({ 
            question: '',
            errorQuestion: '',
            answer: '',
            errorAnswer: '',
          })
          this.props.navigation.dispatch(NavigationActions.navigate({ routeName: 'DeckDetail' }))
        })
    }
  }

  render() {
    const { question, errorQuestion, answer, errorAnswer } = this.state
    const { navigation } = this.props
    const { deckId } = navigation.state.params

    return (
      <View style={Grid.container}>
        <Title>Add Card</Title>
        <CustomTextField
          label='Question'
          value={question}
          error={errorQuestion}
          onChange={(question) => this.setState({ question })}
        />
        <CustomTextField
          label='Answer'
          value={answer}
          error={errorAnswer}
          onChange={(answer) => this.setState({ answer })}
        />
        <CustomSwitch 
          value={this.state.correct}
          label='Is a correct answer?'
          onChange={(value) => this.setState({ correct: value })}
        />
        <PressBtn onPress={() => this.onSubmit(deckId)} label='SAVE' />
      </View>
    )
  }
}

export default connect()(AddCard)