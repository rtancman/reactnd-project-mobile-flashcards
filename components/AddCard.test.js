import React from 'react'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer';
import { mockDeck } from '../fixtures/mocks';
import AddCard from './AddCard'


describe('AddCard Component', () => {
  const middlewares = [thunk]
  const mockStore = configureStore(middlewares)
  let store

  beforeEach(() => {
    store = mockStore({
      decks: {
        [mockDeck.id]: mockDeck
      },
    })
  })

  it('without crashing', async () => {
    const props = {
      store: store,
    }
    const navigation = { 
      navigate: jest.fn(),
      state: {
        params: {
          deckId: mockDeck.id
        }
      },
    }

    const component = renderer.create(
      <Provider {...props}>
        <AddCard navigation={navigation} />
      </Provider>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
})