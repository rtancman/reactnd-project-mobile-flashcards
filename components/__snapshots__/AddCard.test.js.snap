// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCard Component without crashing 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "padding": 20,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#708090",
          "fontSize": 28,
          "marginBottom": 20,
        },
        Object {},
      ]
    }
  >
    Add Card
  </Text>
  <View
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    pointerEvents="auto"
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "borderBottomColor": "rgba(112, 128, 144, 1)",
          "borderBottomWidth": 0.5,
          "height": 64,
          "paddingBottom": 8,
          "paddingTop": 32,
        }
      }
    >
      <View
        style={
          Object {
            "position": "absolute",
            "top": 36,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "rgba(112, 128, 144, 1)",
              "fontSize": 16,
            }
          }
        >
          Question
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          autoCapitalize="sentences"
          autoCorrect={false}
          disableFullscreenUI={true}
          editable={true}
          error=""
          onBlur={[Function]}
          onChange={[Function]}
          onChangeText={[Function]}
          onContentSizeChange={[Function]}
          onFocus={[Function]}
          selectionColor="#C0C0C0"
          style={
            Array [
              Object {
                "flex": 1,
                "margin": 0,
                "padding": 0,
                "top": 2,
              },
              Object {
                "color": "#708090",
                "fontSize": 16,
                "height": 24,
                "textAlign": "left",
              },
              undefined,
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "height": 8,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "flex-start",
              "bottom": 0,
              "left": 0,
              "paddingVertical": 4,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "rgb(213, 0, 0)",
                "fontSize": 0,
                "opacity": 0,
              }
            }
          >
            
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "flex-start",
              "bottom": 0,
              "left": 0,
              "paddingVertical": 4,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#708090",
                "fontSize": 12,
                "opacity": 1,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
  <View
    onResponderRelease={[Function]}
    onStartShouldSetResponder={[Function]}
    pointerEvents="auto"
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "borderBottomColor": "rgba(112, 128, 144, 1)",
          "borderBottomWidth": 0.5,
          "height": 64,
          "paddingBottom": 8,
          "paddingTop": 32,
        }
      }
    >
      <View
        style={
          Object {
            "position": "absolute",
            "top": 36,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "rgba(112, 128, 144, 1)",
              "fontSize": 16,
            }
          }
        >
          Answer
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          autoCapitalize="sentences"
          autoCorrect={false}
          disableFullscreenUI={true}
          editable={true}
          error=""
          onBlur={[Function]}
          onChange={[Function]}
          onChangeText={[Function]}
          onContentSizeChange={[Function]}
          onFocus={[Function]}
          selectionColor="#C0C0C0"
          style={
            Array [
              Object {
                "flex": 1,
                "margin": 0,
                "padding": 0,
                "top": 2,
              },
              Object {
                "color": "#708090",
                "fontSize": 16,
                "height": 24,
                "textAlign": "left",
              },
              undefined,
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "height": 8,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "flex-start",
              "bottom": 0,
              "left": 0,
              "paddingVertical": 4,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "rgb(213, 0, 0)",
                "fontSize": 0,
                "opacity": 0,
              }
            }
          >
            
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "flex-start",
              "bottom": 0,
              "left": 0,
              "paddingVertical": 4,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#708090",
                "fontSize": 12,
                "opacity": 1,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 30,
        "marginTop": 30,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#708090",
            "fontSize": 16,
          },
          Object {
            "flex": 1,
          },
        ]
      }
    >
      Is a correct answer?
    </Text>
    <RCTSwitch
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "height": 31,
            "width": 51,
          },
          Object {
            "flex": 1,
          },
        ]
      }
      value={false}
    />
  </View>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#708090",
        "borderRadius": 7,
        "height": 45,
        "marginBottom": 20,
        "marginLeft": 40,
        "marginRight": 40,
        "opacity": 1,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 22,
          "textAlign": "center",
        }
      }
    >
      
       
      SAVE
    </Text>
  </View>
</View>
`;
